### ISSUE TITLE
Imbalanced metric states with `drop_last=False`

### ISSUE THREAD

**Issue by ananyahjha93** [Lightning-AI/torchmetrics#2170]
## 🐛 Bug

So if I use metrics during eval with drop_last=False, different ranks might end up with metric states which have different lengths.

trivial example:

Metric state is a list

batch_size: 5
ranks: 2
drop_last=False
num_samples in eval: 8

```
rank0: [tensor(0), tensor(1), tensor(2), tensor(3), tensor(4)]
rank1: [tensor(0), tensor(1), tensor(2)]
```

This causes the sync before `compute` to crash.

### Expected behavior

Is it a good idea to add a `before_compute` callback to the Metrics class which allows users to pad/fix metric states before distributed sync is called for the `compute` function?

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source): 1.0.3
- Python & PyTorch Version (e.g., 1.0): 3.10, 2.1

**Comment by ananyahjha93** [comment]
@Borda @SkafteNicki

**Comment by ananyahjha93** [comment]
Also I looked at `gather_all_tensors` function in utilities/distributed.py where it looked like it is supposed to pad tensors to a common size for sync, however, if metric states are a list of tensors, then this gets called on individual tensor elements of the list and not on the whole list.

**Comment by ananyahjha93** [comment]
For now I am using a simple solution which adds a callback type function before sync is called for compute.

```
    def _sync_dist(self, dist_sync_fn: Callable = gather_all_tensors, process_group: Optional[Any] = None) -> None:
        input_dict = {attr: getattr(self, attr) for attr in self._reductions}

        self.before_sync(input_dict)

        for attr, reduction_fn in self._reductions.items():
            # pre-concatenate metric states that are lists to reduce number of all_gather operations
            if reduction_fn == dim_zero_cat and isinstance(input_dict[attr], list) and len(input_dict[attr]) > 1:
                input_dict[attr] = [dim_zero_cat(input_dict[attr])]
        ...
```

```
def before_sync(self, input_dict):
        # get a len of all list type metric states
        state_length = {}
        for key, val in input_dict.items():
            if isinstance(val, list):
                state_length[key] = torch.tensor(len(input_dict[key])).to(input_dict[key][0][0].device)

        # gather list lengths from all ranks
        all_gather_state_lengths = {}
        for key in state_length:
            all_gather_state_lengths[key] = [torch.zeros_like(state_length[key]) for _ in range(torch.distributed.get_world_size())]
            torch.distributed.all_gather(all_gather_state_lengths[key], state_length[key])

        # state_length now contains maximum
        for key, val in all_gather_state_lengths.items():
            state_length[key] = torch.max(torch.stack(all_gather_state_lengths[key]))

        # pad input_dict according to max length on each rank
        # duplicate last element, metric class is designed to handle duplicates
        for key, val in input_dict.items():
            max_len = torch.max(torch.stack(all_gather_state_lengths[key]))
            if isinstance(val, list) and len(val) < max_len:
                diff = max_len - len(val)
                duplicate_element = deepcopy(input_dict[key][-1])
                while diff > 0:
                    input_dict[key].append(duplicate_element)
                    diff -= 1

         return input_dict
```

**Comment by SkafteNicki** [comment]
Hi @ananyahjha93, thanks for raising this issue.
I was actually convinced that this would not be a problem in torchmetrics because we internally would deal with these using the padding, but clearly there is a case here that is not covered. The code for distributed has not been touched in a very long time.
I look into it and see what I can come up with.

**Comment by SkafteNicki** [comment]
@ananyahjha93 just to be sure, how do you initialize the state?
I assume that you use `self.add_state` but in this case what do you set `dist_reduce_fx` to?

**Comment by ananyahjha93** [comment]
@SkafteNicki let me try the cat, the simplest solution never crossed my mind!

this is how I init the states

```
        self.add_state('loglikelihoods', default=[], dist_reduce_fx=None)
        self.add_state('labels', default=[], dist_reduce_fx=None)
```

**Comment by ananyahjha93** [comment]
this custom metric class for in context learning which collects context and continuations across multiple GPUs and computes accuracy.
